crates/transport_iroh/src/lib.rs

```diff
@@ -0,0 +1,656 @@
+    pub fn create() -> DynTransportFactory {
+        Arc::new(Self)
+    }
+}
+
+impl TransportFactory for IrohTransportFactory {
+    fn default_config(&self, config: &mut Config) -> K2Result<()> {
+        config.set_module_config(&IrohTransportModConfig::default())
+    }
+
+    fn validate_config(&self, config: &Config) -> K2Result<()> {
+        let config: IrohTransportModConfig = config.get_module_config()?;
+
+        if let Some(relay) = &config.iroh_transport.relay_url {
+            let relay_url = ::url::Url::parse(relay)
+                .map_err(|err| K2Error::other_src("invalid relay URL", err))?;
+            let uses_tls = relay_url.scheme() == "https";
+            if !&config.iroh_transport.relay_allow_plain_text && !uses_tls {
+                return Err(K2Error::other(format!(
+                    "disallowed plaintext relay url, either specify https or set relay_allow_plain_text to true: {relay_url}"
+                )));
+            }
+        }
+        Ok(())
+    }
+
+    fn create(
+        &self,
+        builder: Arc<Builder>,
+        handler: DynTxHandler,
+    ) -> BoxFut<'static, K2Result<DynTransport>> {
+        Box::pin(async move {
+            let handler = TxImpHnd::new(handler);
+            let config: IrohTransportModConfig =
+                builder.config.get_module_config()?;
+            let imp =
+                IrohTransport::create(config.iroh_transport, handler.clone())
+                    .await?;
+            Ok(DefaultTransport::create(&handler, imp))
+        })
+    }
+}
+
+#[derive(Debug)]
+struct IrohTransport {
+    endpoint: Arc<Endpoint>,
+    handler: Arc<TxImpHnd>,
+    local_url: Arc<RwLock<Option<Url>>>,
+    connections: Arc<RwLock<HashMap<Url, Arc<ConnectionContext>>>>,
+    connection_locks: Arc<Mutex<HashMap<Url, Arc<tokio::sync::Mutex<()>>>>>,
+    watch_addr_task: AbortHandle,
+    accept_task: AbortHandle,
+}
+
+impl Drop for IrohTransport {
+    fn drop(&mut self) {
+        self.watch_addr_task.abort();
+        self.accept_task.abort();
+    }
+}
+
+impl IrohTransport {
+    async fn create(
+        config: IrohTransportConfig,
+        handler: Arc<TxImpHnd>,
+    ) -> K2Result<DynTxImp> {
+        // If a relay server is configured, only use that.
+        // Otherwise use the default relay servers provided by n0.
+        let mut builder = if let Some(relay_url) = config.relay_url {
+            let relay_url =
+                RelayUrl::from_str(&relay_url).map_err(K2Error::other)?;
+            let relay_map = RelayMap::from_iter([relay_url]);
+            Endpoint::empty_builder(RelayMode::Custom(relay_map))
+        } else {
+            Endpoint::empty_builder(RelayMode::Default)
+        };
+        // Set kitsune2 protocol for handling data.
+        builder = builder.alpns(vec![ALPN.to_vec()]);
+
+        let endpoint = builder.bind().await.map_err(|err| {
+            K2Error::other_src("failed to bind iroh endpoint", err)
+        })?;
```

Comment:
Do you know if this is offline friendly? Like will it bind even if it can't reach the relay and keep trying to get a connection?